primitives/runtime/src/offchain/storage_lock.rs

```diff
@@ -0,0 +1,445 @@
+//!
+//! ## Example:
+//!
+//! ```rust
+//! # use crate::offchain::storage::StorageValueRef;
+//! # use codec::{Decode, Encode, Codec};
+//! // in your off-chain worker code
+//!
+//! fn append_to_in_storage_vec<'k, T>(key: &'k [u8], _: T) where T: Encode {
+//!    let mut lock = StorageLock::new(b"x::lock");
+//!    {
+//!         let _guard = lock.spin_lock();
+//!         let acc = StorageValueRef::persistent(key);
+//!         let v: Vec<T> = acc.get::<Vec<T>>().unwrap().unwrap();
+//!         // modify `v` as desired - i.e. perform some heavy computation or side effects that should only be done once.
+//!         acc.set(v);
+//!         // drop `_guard` implicitly at end of scope
+//!    }
+//! }
+//! ```
+
+use crate::offchain::storage::StorageValueRef;
+use codec::{Codec, Decode, Encode};
+use sp_core::offchain::{Duration, Timestamp};
+use sp_io::offchain;
+
+//use frame_system how?
+
+/// Default expiry duration in milliseconds.
+const STORAGE_LOCK_DEFAULT_EXPIRY_DURATION_MS: u64 = 30_000;
+
+/// Snooze duration before attempting to lock again in ms.
+const STORAGE_LOCK_PER_CHECK_ITERATION_SNOOZE: u64 = 100;
+
+/// Lockable item for use with a persisted storage lock.
+///
+/// Bound for an item that has a stateful ordered meaning
+/// without explicitly requiring `Ord` trait in general.
+pub trait Lockable: Sized + Codec + Copy {
+	/// Get the current value of lockable.
+	fn current() -> Self;
+
+	/// Acquire a new deadline based on `Self::current()`
+	fn deadline() -> Self;
+
+	/// Verify the current value of `self` against `deadline`
+	/// to determine if the lock has expired.
+	fn expired(&self, deadline: &Self) -> bool;
+
+	/// Snooze the thread for time determined by `self` and `other`.
+	///
+	/// Only called if not expired just yet.
+	/// Note that the deadline is only passed to allow some optimizations
+	/// for some `L` types.
+	fn snooze(&self, _deadline: &Self) {
+		sp_io::offchain::sleep_until(offchain::timestamp().add(Duration::from_millis(
+			STORAGE_LOCK_PER_CHECK_ITERATION_SNOOZE,
+		)));
+	}
+}
+
+impl Lockable for Timestamp {
+	fn current() -> Self {
+		offchain::timestamp()
+	}
+
+	fn deadline() -> Self {
+		Self::current().add(Duration::from_millis(
+			STORAGE_LOCK_DEFAULT_EXPIRY_DURATION_MS,
+		))
+	}
+
+	fn expired(&self, reference: &Self) -> bool {
+		<Self as Lockable>::current() > *reference
+	}
+
+	fn snooze(&self, deadline: &Self) {
+		let remainder: Duration = self.diff(&deadline);
+		// do not snooze the full duration, but instead snooze max 100ms
+		// it might get unlocked in another thread
+		// consider adding some additive jitter here
+		let snooze = core::cmp::min(remainder.millis(), STORAGE_LOCK_PER_CHECK_ITERATION_SNOOZE);
+		sp_io::offchain::sleep_until(self.add(Duration::from_millis(snooze)));
+	}
+}
+
+/// Lockable based on the current block number and a timestamp based deadline.
+#[derive(Debug, Copy, Clone, Encode, Decode)]
+pub struct BlockAndTime<B>
+where
+	B: BlockNumberTrait,
+{
```

Comment:
```suggestion
#[derive(Debug, Copy, Clone, Encode, Decode)]
pub struct BlockAndTime<B: BlockNumberTrait> {
```

shorter

Reply:
So having had brief look into the code base, it seems that most"simple" bounds are done with the suggested notation, but more complex ones are done with the `where` keyword.

Reply:
I am not sure what our general guidelines about that (@bkchr ?) 

I think that if you have a bound directly on struct then there is no point of using where